### PR TITLE
Modify the dataset code to accept residue position data

### PR DESCRIPTION
**Pull request by cbaakman**
This pull includes several changes, but the most important one is the alterations to the Dataset class. The mapfly option has been removed (advice from Nico) and the dataset no longer needs two chain identifiers. The test_learn test has been modified to see whether this code throws any errors. It succeeds.

Additionally, other tests have been fixed in this pull. It's not up to the reviewer to read all these changes. They have simply been made to get rid of warning messages that appear with each new push.

### CODE REVIEW

File: deeprank/learn/DataSet.py [hunk 1]
```diff
@@ -1179,52 +1032,6 @@ def load_one_molecule(self, fname, mol=None):
         return (np.array(feature).astype(outtype),
                 np.array([target]).astype(outtype))
```
Comment (rgayatri) [inline deeprank/learn/DataSet.py:1033]: 1. Line 949: we would have one target per variant, instead of molecule
2. Line 768: There would be only two values 0.0 and 1.0 so min/max for targets may not be required
3. Line 833-885: We don't have a list of targets. So target ordering, ranking, normalization may not be required

File: deeprank/learn/DataSet.py [hunk 2]
```diff
@@ -535,36 +524,35 @@ def filter(self, molgrp):
...
 
         for cond_name, cond_vals in self.dict_filter.items():
 
-            try:
-                val = molgrp['targets/' + cond_name][()]
-            except KeyError:
-                warnings.warn(f'Filter {cond_name} not found for mol '
-                              f'{molgrp.name}')
+            logger.debug("filter with condition {}: {}".format(cond_name, cond_vals))
+
```
Comment (rgayatri) [inline deeprank/learn/DataSet.py:543]: cond_name and cond_vals example- 'Benign':'0.0' or 'Pathogenic': '1.0' ?
If we plan on additional class labels (I'll ask Bas) we can add.

File: deeprank/learn/DataSet.py [hunk 3]
```diff
@@ -527,6 +493,29 @@ def _select_pdb(self, mol_names):
...
+                        search_index = found_index + len(operator_string)
+                    else:
+                        return_string = return_string[:found_index] + inserting_string + return_string[found_index:]
+
+                        search_index = found_index + len(inserting_string) + len(operator_string)
+                else:
+                    # not found
+                    break
+
```
Comment (rgayatri) [inline deeprank/learn/DataSet.py:516]: What is this function useful for?

File: deeprank/learn/DataSet.py [hunk 4]
```diff
@@ -25,8 +25,7 @@
 class DataSet():
 
     def __init__(self, train_database, valid_database=None, test_database=None,
-                 chain1='A', chain2='B',
-                 mapfly=True, grid_info=None,
+                 grid_info=None,
                  use_rotation=None,
                  select_feature='all', select_target='DOCKQ',
```
Comment (rgayatri) [inline deeprank/learn/DataSet.py:30]: We probably won't have users deciding what target they want, as the program will predict.

File: deeprank/learn/DataSet.py (hunk 1 above)
Comment (cbaakman) [reply deeprank/learn/DataSet.py:1033]: I've changed the name of molecule to variant. (1) About (2) and (3) I'm not sure. Can those unnecessary steps do harm?

Comment (rgayatri) [reply deeprank/learn/DataSet.py:1033]: They might throw errors I suppose, maybe semantic?

Comment (rgayatri) [reply deeprank/learn/DataSet.py:1033]: Or we could set normalize_targets to False, target_ordering to None

File: deeprank/learn/DataSet.py (hunk 3 above)
Comment (cbaakman) [reply deeprank/learn/DataSet.py:516]: In the original deeprank, one of the authors appears to have made a "filter" function. In this filter function, the variable name "val" appears to be placed before the operators: '>=', '<=', '==', '>', '<'. When I was testing this function however, I bumped into a bug, caused by the operator string not getting detected properly, resulting in an error. That's why I made the function "_insert_before_operators". The purpose of this function is to make sure that the string "val" gets inserted properly before any of the operators. I'm adding a docstring..

File: deeprank/learn/DataSet.py (hunk 2 above)
Comment (cbaakman) [reply deeprank/learn/DataSet.py:552]: I see that right now, the filter function uses the operators '>=', '<=', '==', '>', '<'. So '!=' is missing. Maybe we still need to change it then..

File: deeprank/learn/DataSet.py (hunk 1 above)
Comment (rgayatri) [reply deeprank/learn/DataSet.py:1042]: We can set normalize_targets= False and target_ordering=None
This should be fine I suppose. 
I'll change the associated readme file for this.

File: deeprank/learn/DataSet.py (hunk 4 above)
Comment (cbaakman) [reply deeprank/learn/DataSet.py:30]: It's used in the methods "get_norm" and "load_one_variant" though. Wouldn't it be better to set it to something else?
Like: "class".

Comment (rgayatri) [reply deeprank/learn/DataSet.py:30]: Yes, but if we set select_target='class' but we must make sure this option is not modifiable for the user. So yes, I'll make sure readme file clarifies this. 
normalize_target must always be False.

Also, in line 892 and 893, since we have 0.0 and 1.0 when program tries to normalize- it will try dividing 0 by 1 so this might be problematic when both min and max values are 0.

Comment (cbaakman) [reply deeprank/learn/DataSet.py:30]: I removed the normalize target feature in this commit: https://github.com/DeepRank/deeprank/commit/fb5f304b9a0c9f508db78f9233ea27083fb8538c

File: deeprank/learn/DataSet.py (hunk 2 above)
Comment (rgayatri) [reply deeprank/learn/DataSet.py:545]: Since the values we have are binary, may be <=, >=, >, < may not make sense. 
!= can be added, yes

File: deeprank/learn/DataSet.py (hunk 4 above)
Comment (rgayatri) [reply deeprank/learn/DataSet.py:30]: great

File: deeprank/learn/NeuralNet.py [hunk 5]
```diff
@@ -191,8 +191,6 @@ def __init__(self, data_set, model,
         elif self.task == 'class':
```
Comment (rgayatri) [inline deeprank/learn/NeuralNet.py:191]: chain1 chain2 must be change
Line 38- save_hitrate= False?

File: deeprank/learn/NeuralNet.py [hunk 6]
```diff
@@ -991,6 +972,7 @@ def _plot_scatter_reg(self, figname):
         ax.set_ylabel('Predictions')
 
         values = np.append(xvalues, yvalues)
+
         border = 0.1 * (values.max() - values.min())
         ax.plot([values.min() - border, values.max() + border],
                 [values.min() - border, values.max() + border])
```
Comment (rgayatri) [inline deeprank/learn/NeuralNet.py:978]: Functions _plot_hit_rate _compute_hitrate and _get_relevance use iRMSD values that we do not have.
I think we can ask Li about usability of these functions.

Comment (cbaakman) [reply deeprank/learn/NeuralNet.py:978]: **about plot_hit_rate:** this method is called in the "NeuralNet.test" function, that we do not call

**about compute_hitrate**: this is commented out in deeprank's code

**about get_relevance**: we don't use this in our version. The developers adviced me to set "NeuralNet.save_hitrate" to False, after getting an error message.

File: deeprank/learn/NeuralNet.py (hunk 5 above)
Comment (cbaakman) [reply deeprank/learn/NeuralNet.py:184]: I changed it: https://github.com/DeepRank/deeprank/commit/0768e89e68c411b40fa6f031fa8fa9998119d0ec

File: test/test_learn.py [hunk 7]
```diff
@@ -1,260 +1,78 @@
...
 
-        # adress of the database
-        train_database = '1ak4.hdf5'
-        if not os.path.isfile(train_database):
-            raise FileNotFoundError(
-                'Database %s not found. Make sure to run test_generate before',
-                train_database)
+        # data_augmentation has been set to a high number, so that
+        # the train, valid and test set can be large enough.
+        data_generator = DataGenerator([mutant], data_augmentation=50,
```
Comment (rgayatri) [inline test/test_learn.py:48]: How is the number for data augmentation arrived at?

File: deeprank/learn/NeuralNet.py [hunk 8]
```diff
@@ -71,11 +69,7 @@ def __init__(self, data_set, model,
                 Only applicable on 'class' task.
```
Comment (rgayatri) [inline deeprank/learn/NeuralNet.py:69]: What would class weights be?

File: deeprank/learn/NeuralNet.py [hunk 9]
```diff
@@ -31,11 +31,9 @@ def __init__(self, data_set, model,
                  model_type='3d', proj2d=0, task='reg',
                  class_weights = None,
                  pretrained_model=None,
-                 chain1='A',
-                 chain2='B',
                  cuda=False, ngpu=0,
                  plot=True,
```
Comment (rgayatri) [inline deeprank/learn/NeuralNet.py:35]: Would we have enough data to plot?
Input in our case will be one variant, PSSM, and one structure file. So I would expect one prediction for one variant? 
Plots would make sense we have a dataset of variants.

File: test/test_atomic_features.py [hunk 10]
```diff
@@ -16,9 +16,9 @@ def test_atomic_haddock():
         # - comment the first line (E0=1)
```
Comment (rgayatri) [inline test/test_atomic_features.py:16]: Is this test just to see if the code functions?

File: test/test_learn.py [hunk 11]
```diff
@@ -1,260 +1,78 @@
...
-            save_model='all')
+        ok_(len(dataset) > 0)
+        ok_(dataset[0] is not None)
 
-    @unittest.skipIf(skip, "Torch fails on Travis")
-    @staticmethod
-    def test_learn_3d_class():
-        """Use a 3D CNN for regularization."""
+        net_output_dir_path = os.path.join(work_dir_path, 'net-output')
+        neural_net = NeuralNet(dataset, cnn_reg, model_type='3d',task='reg', save_hitrate=False,
```
Comment (rgayatri) [inline test/test_learn.py:68]: Here the 'dataset' is features associated with one variant, isn't it?

Comment (cbaakman) [reply test/test_learn.py:68]: Yes, one variant. However, I used data augmentation to make 50 orientations of the variant's structure. (lazy) All this tests is that the NeuralNet.train method doesn't throw an error. The test doesn't look at the actual results.

Do you wish to extend this test somehow?

File: test/test_atomic_features.py (hunk 10 above)
Comment (cbaakman) [reply test/test_atomic_features.py:16]: Yes, all this tests is that the NeuralNet.train method doesn't throw an error. The test doesn't look at the actual results.

Do you wish to extend this test somehow?

File: test/test_learn.py (hunk 7 above)
Comment (cbaakman) [reply test/test_learn.py:48]: I needed a set of input data large enough to train the network on. So that there's room for a train, validation and test set. At the same time, the larger the set, the longer the test takes. So I decided on 50. It could be smaller I guess.

File: deeprank/learn/DataSet.py (hunk 2 above)
Comment (cbaakman) [reply deeprank/learn/DataSet.py:545]: I added !=

File: deeprank/learn/NeuralNet.py (hunk 8 above)
Comment (cbaakman) [reply deeprank/learn/NeuralNet.py:69]: I guess this should depend on the frequency of variants that occur of each class. Low for benign, because it's common? High for disease, because it's rare?

See also: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

File: deeprank/learn/NeuralNet.py (hunk 9 above)
Comment (cbaakman) [reply deeprank/learn/NeuralNet.py:35]: The data_set argument can take one or more hdf5 files as input. Each hdf5 file can contain multiple variants. I could test this also, if you like.

File: test/test_atomic_features.py (hunk 10 above)
Comment (cbaakman) [reply test/test_atomic_features.py:16]: I did not create this test, I just changed the path names so that the test will run properly from the root directory.

It seems to run the AtomicFeature class and it does some assertions on the output. It checks whether the output values are close to the expected values.

File: test/test_learn.py (hunk 7 above)
Comment (rgayatri) [reply test/test_learn.py:48]: Ok, sure. I think we would have some idea on this once we have data from BioProdict.

File: deeprank/learn/NeuralNet.py (hunk 8 above)
Comment (rgayatri) [reply deeprank/learn/NeuralNet.py:69]: So in the dataset from BioProdict, we have more benign variants evidently:
Benign- 54,138
Pathogenic- 43,460
Although 40% of the data has pathogenic labels, class weights can be higher for it and lower for benign.

File: deeprank/learn/NeuralNet.py (hunk 9 above)
Comment (rgayatri) [reply deeprank/learn/NeuralNet.py:35]: ok, we can keep this and wait for data from BioProdict.

File: test/test_atomic_features.py (hunk 10 above)
Comment (rgayatri) [reply test/test_atomic_features.py:16]: Oh alright. I see it does assertions on the output

File: test/test_learn.py (hunk 11 above)
Comment (rgayatri) [reply test/test_learn.py:68]: Since the test doesn't look at actual results, I think the existing code suffices.

File: deeprank/learn/NeuralNet.py (hunk 8 above)
Comment (cbaakman) [reply deeprank/learn/NeuralNet.py:69]: We can sample the data to create a subset that has 50% benign and 50% pathogenic.

File: deeprank/learn/DataSet.py [hunk 12]
```diff
@@ -25,11 +25,10 @@
...
-                 normalize_features=True, normalize_targets=True,
+                 normalize_features=True,
                  target_ordering=None,
                  dict_filter=None, pair_chain_feature=None,
```
[the remark this answers is not in the document]
Comment (cbaakman) [reply deeprank/learn/DataSet.py:33]: In the docstring, they mentioned that clip_features could be needed for native complexes where the coulomb feature might be too large. For the efro project, we use the coulomb feature also.

How about we keep this optional and set it to False by default?

[the remark this answers is not in the document]
Comment (rgayatri) [reply deeprank/learn/DataSet.py:33]: Yes, clip_features= False by default. It seems it removes features larger than the grid.

File: deeprank/learn/DataSet.py [hunk 13]
```diff
@@ -1017,20 +876,6 @@ def backtransform_target(self, data):
         data += self.target_min
```
Comment (rgayatri) [inline deeprank/learn/DataSet.py:876]: Line 846 onwards will throw error/warning if we test with actual data.
If we set target_ordering= None, the program tries to assign lower_list, higher_list or NA_list. So it tries to assign order based on values in 'IRMSD', 'LRMSD', 'HADDOCK'. And when these are unidentified it assigns lower. 

Should we keep this section?

File: deeprank/learn/DataSet.py [hunk 14]
```diff
@@ -527,44 +490,75 @@ def _select_pdb(self, mol_names):
...
-                        o, 'val' + o)
+                new_cond_vals = DataSet._insert_before_operators(cond_vals, 'val')
                 if not eval(new_cond_vals):
                     return False
             else:
-                raise ValueError(
-                    'Conditions not supported', cond_vals)
+                raise ValueError("Conditions not supported", cond_vals)
 
         return True
```
Comment (rgayatri) [inline deeprank/learn/DataSet.py:564]: Line 584 chain_tags can be removed.
Lines 629:632  need to be changed.

Comment (cbaakman) [reply deeprank/learn/DataSet.py:550]: I made a change here: https://github.com/DeepRank/deeprank/commit/fe5422da5595400b32237d9e6633f5d4fee48e74

File: deeprank/learn/DataSet.py (hunk 13 above)
Comment (rgayatri) [reply deeprank/learn/DataSet.py:818]: Coos did you check this one?

File: deeprank/learn/NeuralNet.py (hunk 6 above)
Comment (rgayatri) [reply deeprank/learn/NeuralNet.py:966]: Thanks Coos.
I still find 'compute_hitrate' and 'get_relevance' included in the current version on NeuralNet.py. 
Is there another version where these are not used?

File: deeprank/learn/DataSet.py (hunk 14 above)
Comment (rgayatri) [reply deeprank/learn/DataSet.py:550]: Thanks Coos

File: deeprank/learn/NeuralNet.py (hunk 6 above)
Comment (cbaakman) [reply deeprank/learn/NeuralNet.py:966]: Indeed the _compute_hitrate method is declared, but it's not used anywhere. We might as well remove it. The original call of this function is commented out at line 719.
For the get_relevance method it's about the same story. It's declared, but not used, since we set save_hitrate to False.

Do you want me to remove the two methods completely?

File: deeprank/learn/DataSet.py (hunk 13 above)
Comment (cbaakman) [reply deeprank/learn/DataSet.py:818]: It looks like target_ordering is a class field that isn't used anywhere. It's only set in the _get_target_ordering method, but the value has no effect. We could just remove it, otherwise we need to implement an application for it ourselves.

Do we even need target_ordering if there's only one target?

Comment (rgayatri) [reply deeprank/learn/DataSet.py:818]: Yes, that was exactly my concern in my very first comment. 
We can remove target_ordering, as there is only one target.

File: deeprank/learn/NeuralNet.py (hunk 6 above)
Comment (rgayatri) [reply deeprank/learn/NeuralNet.py:966]: I think plotting is something we need to check once we have all the data from Bioprodict. The plots would make sense when you have a dataset of variants. 
Plotting functions can be retained. The unused _compute_hitrate and _get_relevance could be removed? These functions also use target_ordering

Comment (cbaakman) [reply deeprank/learn/NeuralNet.py:948]: I removed the two methods in this commit: https://github.com/DeepRank/deeprank/commit/01d0f9f8fdd7a93754063b496bbeb5a603917efc I left the hitrate plotting method in, though we cannot use it right now.

File: deeprank/learn/DataSet.py (hunk 13 above)
Comment (cbaakman) [reply deeprank/learn/DataSet.py:784]: I removed the target_ordering here: https://github.com/DeepRank/deeprank/commit/01d0f9f8fdd7a93754063b496bbeb5a603917efc

Comment (rgayatri) [reply deeprank/learn/DataSet.py:784]: Thanks coos

**Review by rgayatri — COMMENTED** [review]
Thank you Coos.
I went through all the scripts multiple times. 
If you do not have any further changes to make shall I go ahead and merge the pull request?

**Comment by cbaakman** [comment]
Seems a good idea to me. If we would want to make changes in the future, we could always open up a new pull.
Thank you!
